core/lib/instrument_api.h

```diff
@@ -779,6 +779,22 @@ DR_API
  * in failure to clean up the right resources, and at process exit
  * time it may return NULL.
  *
+ * On Linux, SYS_execve may or may not result in a thread exit event.
```

Comment:
`cleanup_after_vfork_execve` cleans up dead threads that previously called `SYS_execve` and calls their thread exit callbacks in process (the "may" case).

If `cleanup_after_vfork_execve` ends up never being called (e.g. the program doesn't fork or create threads anymore), then the thread exit callbacks are not called (the "may not" case).

Reply:
Always raising the thread exit event could be more convenient for clients, they wouldn't have to additionally setup pre-`SYS_execve` callbacks in that case.

Reply:
If you would like to change how the thread exits operate for execve and fork that sounds good if it makes things cleaner.  I think breaking compatibility in a minor way vs old behavior will be ok here (we would document it of course).  One problem for execve is that doing things pre-syscall runs the risk of the syscall failing if the predictions for whether it will succeed are off.

Do you want to merge these docs changes here in and separately take a stab at changing the DR behavior?

Reply:
Clarifying what I mean about the syscall failing: if we think SYS_execve will succeed and go exit all the threads, and then the syscall fails, we can't easily recover.  If our prediction of success is good enough we can probably live with that.

Reply:
Yeah, the thread cleanup (including the associated thread exit event) would have to happen somewhere after the `SYS_execve` (if it was successful), but I'm not sure what exactly should trigger it and what should happen if the thread calling `SYS_execve` is the only thread in the program.

So, I'd prefer to merge the documentation changes first.

Reply:
SGTM